File: php/test/DiamondIterativeTest.php
<?php


use PHPUnit\Framework\TestCase;

/**
 * These test cases can be used to test-drive a solution to the diamond kata, in an interative manner.
 * Enable only one test case for "Diamond A", one for "Diamond B" and one for "Diamond C" at any one time.
 * The idea is that you iterate towards a full solution.
 *
 * Instructions:
 * 1. Make the first test for Diamond A (which is failing) pass
 * 2. Remove the 'markTestSkipped' on the next test case. Make it pass too.
 * 3. Replace the 'markTestSkipped' on that test case, (or delete it), and instead enable the following test case
 * 4. keep doing this, always having exactly one enabled test case for each diamond, (A, B, and C) until you've got to the end of the file.
 * At that point you should have a working implementation.
 */
class DiamondIterativeTest extends TestCase
{

    /**
     * @test
     */
    public function DiamondA_has_one_line_containing_A()
    {
        $this->assertEquals("A", Diamond::print('A'));
    }

    /**
     * @test
     */
    public function DiamondB_prints_a_char_sequence_for_the_top_half()
    {
        $this->markTestSkipped();
        $this->assertEquals("AB", Diamond::print('B'));
    }

    /**
     * @test
     */
    public function DiamondB_prints_a_char_sequence_for_the_whole_diamond()
    {
        $this->markTestSkipped();
        $this->assertEquals("ABA", Diamond::print('B'));
    }

    /**
     * @test
     */
    public function DiamondC_prints_a_char_sequence_for_the_whole_diamond()
    {
        $this->markTestSkipped();
        $this->assertEquals("ABCBA", Diamond::print('C'));
    }

    /**
     * @test
     */
    public function DiamondB_has_newlines()
    {
        $this->markTestSkipped();
        $this->assertEquals("A\nB\nA", Diamond::print('B'));
    }

    /**
     * @test
     */
    public function DiamondB_has_indentation_on_first_and_last_rows()
    {
        $this->markTestSkipped();
        $this->assertEquals(" A\nB\n A", Diamond::print('B'));
    }

    /**
     * @test
     */
    public function DiamondC_has_indentation_on_all_rows_except_middle()
    {
        $this->markTestSkipped();
        $this->assertEquals("  A\n" .
                            " B\n" .
                            "C\n" .
                            " B\n" .
                            "  A", Diamond::print('C'));
    }

    /**
     * @test
     */
    public function DiamondB_duplicates_B_on_middle_row()
    {
        $this->markTestSkipped();
        $this->assertEquals(" A\nBB\n A", Diamond::print('B'));
    }

    /**
     * @test
     */
    public function DiamondC_has_duplicates_on_middle_rows()
    {
        $this->markTestSkipped();
        $this->assertEquals("  A\n" .
                            " BB\n" .
                            "CC\n" .
                            " BB\n" .
                            "  A", Diamond::print('C'));
    }

    /**
     * @test
     */
    public function DiamondB_has_a_space_between_the_Bs()
    {
        $this->assertEquals(" A\nB B\n A", Diamond::print('B'));
    }

    /**
     * @test
     */
    public function DiamondC_has_between_spaces_on_all_rows_except_first_and_last()
    {
        $this->assertEquals("  A\n" .
                            " B B\n" .
                            "C   C\n" .
                            " B B\n" .
                            "  A", Diamond::print('C'));
    }

    /**
     * @test
     */
    public function DiamondD_is_correct_in_all_aspects() {
        $this->assertEquals("   A\n" .
                            "  B B\n" .
                            " C   C\n" .
                            "D     D\n" .
                            " C   C\n" .
                            "  B B\n" .
                            "   A", Diamond::print('D'));
    }
}

File: csharp/DiamondKata.Tests/DiamondIterativeTest.cs
using Xunit;

namespace DiamondKata.Tests.Iterative;

/// <summary>
///     <para>These test cases can be used to test-drive a solution to the diamond kata, in an iterative manner.</para>
///     <para>Enable only one test case at a time.</para>
///     <para>The idea is that you iterate towards a full solution.</para>
///     <para>Instructions:</para>
///     <list type="number">
///         <item>Make the first test (which is failing) pass.</item>
///         <item>Remove the 'Skip' property on the next test case. Make it pass too.</item>
///         <item>
///             Replace the 'Skip' property on that test case, (or delete it), and instead enable the following test case.
///         </item>
///         <item>
///             Keep doing this, always having exactly one enabled test case at a time until you've got to the end of the
///             file.
///         </item>
///     </list>
///     <para>At that point you should have a working implementation.</para>
/// </summary>
public class DiamondShould
{
    [Fact]
    public void Print_smallest_one_with_only_A_character()
        => Assert.Equal("A", Diamond.Print('A'));

    [Fact(Skip = "1. Make it pass 2. Skip it again or delete it and go to next test")]
    public void Give_character_sequence_when_B_is_supplied()
        => Assert.Equal("AB", Diamond.Print('B'));

    [Fact(Skip = "1. Make it pass 2. Skip it again or delete it and go to next test")]
    public void Repeat_characters_when_B_is_supplied()
        => Assert.Equal("ABB", Diamond.Print('B'));

    [Fact(Skip = "1. Make it pass 2. Skip it again or delete it and go to next test")]
    public void Have_separate_lines_when_B_is_supplied()
        => Assert.Equal(
            "A\n" +
            "BB\n",
            Diamond.Print('B'));

    [Fact(Skip = "1. Make it pass 2. Skip it again or delete it and go to next test")]
    public void Indent_characters_when_B_is_supplied()
        => Assert.Equal(
            " A\n" +
            "BB\n",
            Diamond.Print('B'));

    [Fact(Skip = "1. Make it pass 2. Skip it again or delete it and go to next test")]
    public void Separate_characters_with_inter_character_spaces_when_B_is_supplied()
        => Assert.Equal(
            " A\n" +
            "B B\n",
            Diamond.Print('B'));

    [Fact(Skip = "1. Make it pass 2. Skip it again or delete it and go to next test")]
    public void Print_the_bottom_half_by_symmetry()
        => Assert.Equal(
            " A\n" +
            "B B\n" +
            " A",
            Diamond.Print('B'));

    [Fact(Skip = "1. Make it pass 2. Congrats, you have a working solution")]
    public void Print_with_supplied_letter_at_the_widest_point()
        => Assert.Equal(
            "    A\n" +
            "   B B\n" +
            "  C   C\n" +
            " D     D\n" +
            "E       E\n" +
            " D     D\n" +
            "  C   C\n" +
            "   B B\n" +
            "    A",
            Diamond.Print('E'));
}